model.py

```diff
@@ -411,3 +423,35 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None, return_logi
             return idx, torch.cat(logits_list, dim=1)
         
         return idx
+
+    # *yoink*: https://github.com/tcapelle/llama2.c/blob/af8597ca6ac0c74142951c44b8d6926e170d791a/model.py#L428
+    # the following metric computations require that the model be run in eval mode, then the respective layer outputs
+    # will be saved in order to compute the metric of the latest input passed through the model
+    def compute_attention_metrics(self) -> Tuple[List[float], List[float]]:
+        "compute the max inf norm and kurtosis of the attention outputs"
+        outputs = [b.attention.output for b in self.layers]
+        k = [kurtosis(o.flatten().half()) for o in outputs]
+        inf_norm = [o.abs().max().item() for o in outputs]
+        return inf_norm, k
+
+    def compute_ffn_metrics(self) -> Tuple[List[float], List[float]]:
+        "compute the max inf norm and kurtosis of the ffn outputs"
+        outputs = [b.feed_forward.output for b in self.layers]
+        k = [kurtosis(o.flatten().half()) for o in outputs]
+        inf_norm = [o.abs().max().item() for o in outputs]
+        return inf_norm, k
+
+    def compute_softmax_metrics(self) -> torch.Tensor:
```

Comment:
In the future, we should have pure functions as a `@staticmethod` and add PyTests.